// @sap-ux/axios-extension/dist/abap/adt-catalog/adt-catalog-service.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdtCatalogService = void 0;
const axios_1 = require("axios");
const adt_schema_store_1 = require("./adt-schema-store");
const fast_xml_parser_1 = require("fast-xml-parser");
/**
 * Adt Catalog Service implementation fetches the
 * Adt service specification for a given ADT service.
 */
class AdtCatalogService extends axios_1.Axios {
    constructor() {
        super(...arguments);
        // Cache of fetched discovery schema
        this.schemaStore = new adt_schema_store_1.AdtSchemaStore();
    }
    /**
     * Adt Catalog Service which fetches the Adt service
     * specification for a given ADT service.
     *
     * @param adtCategory Adt service Id
     * @returns Service schema of the input Adt service
     */
    getServiceDefinition(adtCategory) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.checkOrLoadAdtDiscoverySchema();
            // Find the schema for the input service url path
            const serviceSchema = this.schemaStore.getAdtCollection(adtCategory);
            const isValidSchema = this.validateServiceSchema(adtCategory, serviceSchema);
            if (isValidSchema) {
                return serviceSchema;
            }
            else {
                this.log.warn('Invalid Discovery Schema');
                return null;
            }
        });
    }
    /**
     *
     * @param adtCategory adtCategory
     * @param serviceSchema serviceSchema
     * @returns boolean boolean result of schema validity
     */
    validateServiceSchema(adtCategory, serviceSchema) {
        if (!serviceSchema) {
            this.log.warn(`Schema Not Found: ${adtCategory.term} - ${adtCategory.scheme}`);
            return false;
        }
        else if (!serviceSchema.href) {
            this.log.warn(`Empty href in schema: ${adtCategory.term} - ${adtCategory.scheme}`);
            return false;
        }
        else {
            return true;
        }
    }
    /**
     * Check if discover schema is in the local cache. If not, fetch it by
     * calling discover service request.
     *
     * @returns Promise<void>
     */
    checkOrLoadAdtDiscoverySchema() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.schemaStore.isAdtSchemaEmpty()) {
                return;
            }
            const response = yield this.get('', {
                headers: {
                    Accept: 'application/*'
                }
            });
            const schemaData = this.parseAdtSchemaData(response.data);
            this.schemaStore.updateSchemaData(schemaData);
        });
    }
    /**
     *
     * @param xml Raw XML response data of discovery service request
     * @returns Discovery schema data object
     */
    parseAdtSchemaData(xml) {
        if (fast_xml_parser_1.XMLValidator.validate(xml) !== true) {
            return null;
        }
        const options = {
            attributeNamePrefix: '',
            ignoreAttributes: false,
            ignoreNameSpace: true,
            parseAttributeValue: true,
            removeNSPrefix: true
        };
        const parser = new fast_xml_parser_1.XMLParser(options);
        const parsed = parser.parse(xml, true);
        if (parsed.service) {
            return parsed;
        }
        else {
            return null;
        }
    }
}
exports.AdtCatalogService = AdtCatalogService;
// Discovery service url provided by ADT team
AdtCatalogService.ADT_DISCOVERY_SERVICE_PATH = '/sap/bc/adt/discovery';
