// @sap-ux/axios-extension/dist/abap/adt-catalog/adt-schema-store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdtSchemaStore = void 0;
/**
 * This class stores the ADT schema fetched by calling ADT discovery service.
 */
class AdtSchemaStore {
    /**
     * Given the ID of a particular ADT service, return the schema of this service.
     *
     * @param adtCategory ADT service serves as unique id of a service schema
     * @returns Schema of an ADT service
     */
    getAdtCollection(adtCategory) {
        if (!this.adtSchema) {
            return null;
        }
        const id = this.serializeAdtCategory(adtCategory);
        return this.adtSchema[id];
    }
    /**
     * Convert the raw ADT schema data structure toP
     *
     * key-value map for fast access.
     *
     * @param schemaData Raw ADT schema data structure that matches the XML schema
     *                   received from backend
     */
    updateSchemaData(schemaData) {
        if (schemaData) {
            this.adtSchema = {};
            const workspaces = schemaData.service.workspace;
            for (const workspace of workspaces) {
                const collections = workspace.collection;
                if (!collections) {
                    continue;
                }
                if (Array.isArray(collections)) {
                    collections.forEach((collection) => {
                        collection.workspaceTitle = workspace.title;
                        const id = this.serializeAdtCategory(collection.category);
                        this.adtSchema[id] = collection;
                    });
                }
                else {
                    const collection = collections;
                    const id = this.serializeAdtCategory(collection.category);
                    this.adtSchema[id] = collections;
                }
            }
        }
    }
    /**
     * Check if an schema has been loaded and cached.
     *
     * @returns boolean isAdtSchemaEmpty
     */
    isAdtSchemaEmpty() {
        return !this.adtSchema;
    }
    /**
     *
     * @param adtCategory adtCategory
     * @returns string serializeAdtCategory
     */
    serializeAdtCategory(adtCategory) {
        return `${adtCategory.scheme},${adtCategory.term}`;
    }
}
exports.AdtSchemaStore = AdtSchemaStore;
