// @sap/ux-specification/dist/src/sync/v2/import/app/ovpProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2OvpProvider = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const appProvider_1 = require("../../../common/appProvider");
const common_1 = require("../common");
const overviewPage_1 = require("../pages/overviewPage");
const application_1 = require("../../application");
const common_2 = require("../../../common");
class V2OvpProvider extends appProvider_1.AppProvider {
    constructor(importParameters) {
        const pages = {};
        const sapOvp = importParameters.manifest[ux_specification_types_1.ManifestSection.ovp];
        const appSettings = new application_1.AppSettings();
        const schemaFile = (0, common_2.getSchemaFilePath)(ux_specification_types_1.SchemaType.Application);
        (0, common_1.transferSettingsOfObject)(appSettings, importParameters.manifest, importParameters.appSchemas[schemaFile.filename]['definitions']['AppSettings'], undefined);
        const jsonSchema = Object.keys(importParameters.appSchemas).find((element) => element.includes(ux_specification_types_1.PageTypeV2.OverviewPage));
        const parameters = {
            manifest: importParameters.manifest,
            jsonSchema: importParameters.appSchemas[jsonSchema],
            logger: importParameters.logger
        };
        const page = {
            pageType: ux_specification_types_1.PageTypeV2.OverviewPage,
            config: (0, overviewPage_1.createOverviewPageConfig)(parameters),
            entitySet: sapOvp.globalFilterEntitySet
        };
        const id = (0, common_2.generatePageId)(page.pageType, page.entitySet);
        pages[id] = page;
        super(importParameters.manifest, importParameters.appSchemas, id, pages, ux_specification_types_1.FioriElementsVersion.v2, appSettings, importParameters.logger);
    }
}
exports.V2OvpProvider = V2OvpProvider;
