// @sap-ux/logger/dist/winston-logger/vscode-output-channel-transport.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VSCodeTransport = void 0;
const winston_transport_1 = __importDefault(require("winston-transport"));
/**
 * Winston transport for output to VSCode channel
 */
class VSCodeTransport extends winston_transport_1.default {
    constructor(options) {
        super(options);
        this.winstonToVSCodeMap = new Map([
            ['error', 'error'],
            ['warn', 'warn'],
            ['info', 'info'],
            ['verbose', 'debug'],
            ['debug', 'debug'],
            ['silly', 'trace']
        ]);
        this.channel = getVSCodeInstance().window.createOutputChannel(options.channelName, { log: true });
    }
    log(data, callback) {
        setImmediate(() => {
            var _a;
            const logFunction = (_a = this.winstonToVSCodeMap.get(data.level)) !== null && _a !== void 0 ? _a : 'info';
            this.channel[logFunction](data.message);
        });
        callback();
    }
    show() {
        this.channel.show();
    }
}
exports.VSCodeTransport = VSCodeTransport;
function getVSCodeInstance() {
    return require('vscode');
}
