// @sap/ux-specification/dist/src/sync/v4/export/controls/ObjectPageTableColumn.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableCustomColumnOP = exports.PositionOP = exports.ObjectPageTableColumnAction = exports.ObjectPageTableColumn = exports.ActionAfterExecutionConfigurationOP = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
class ActionAfterExecutionConfigurationOP {
}
exports.ActionAfterExecutionConfigurationOP = ActionAfterExecutionConfigurationOP;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getColumnsPathForOP)(pageName, manifest, sectionId, columnKey) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfigurationOP.prototype, "navigateToInstance", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getColumnsPathForOP)(pageName, manifest, sectionId, columnKey) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfigurationOP.prototype, "enableAutoScroll", void 0);
class ObjectPageTableColumn {
}
exports.ObjectPageTableColumn = ObjectPageTableColumn;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ObjectPageTableColumn.prototype, "width", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.85.0'
    })
], ObjectPageTableColumn.prototype, "availability", void 0);
class ObjectPageTableColumnAction extends ObjectPageTableColumn {
}
exports.ObjectPageTableColumnAction = ObjectPageTableColumnAction;
class PositionOP {
}
exports.PositionOP = PositionOP;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getColumnsPathForOP)(pageName, manifest, sectionId, columnKey) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], PositionOP.prototype, "anchor", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getColumnsPathForOP)(pageName, manifest, sectionId, columnKey) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], PositionOP.prototype, "placement", void 0);
class TableCustomColumnOP {
}
exports.TableCustomColumnOP = TableCustomColumnOP;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableCustomColumnOP.prototype, "position", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableCustomColumnOP.prototype, "header", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableCustomColumnOP.prototype, "width", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableCustomColumnOP.prototype, "template", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.86.0'
    })
], TableCustomColumnOP.prototype, "horizontalAlign", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableCustomColumnOP.prototype, "availability", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.85.0'
    })
], TableCustomColumnOP.prototype, "properties", void 0);
