// @sap/ux-specification/dist/src/sync/v4/export/controls/ObjectPageHeaderSection.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeaderSections = exports.ObjectPageHeaderSectionAddress = exports.ObjectPageHeaderSectionContact = exports.ObjectPageHeaderSectionDataPoint = exports.ObjectPageHeaderSectionChart = exports.ObjectPageHeaderSectionForm = exports.CommonHeaderFacetSettings = exports.FlexSettings = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
class FlexSettings {
}
exports.FlexSettings = FlexSettings;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId) => (0, application_1.getHeaderFacetSettingPath)(pageName, manifest, sectionId) + '/flexSettings'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.81.0'
    })
], FlexSettings.prototype, "designtime", void 0);
class CommonHeaderFacetSettings {
}
exports.CommonHeaderFacetSettings = CommonHeaderFacetSettings;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getHeaderFacetSettingPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.81.0'
    })
], CommonHeaderFacetSettings.prototype, "stashed", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getHeaderFacetSettingPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.81.0'
    })
], CommonHeaderFacetSettings.prototype, "flexSettings", void 0);
class ObjectPageHeaderSectionForm extends CommonHeaderFacetSettings {
}
exports.ObjectPageHeaderSectionForm = ObjectPageHeaderSectionForm;
class ObjectPageHeaderSectionChart extends CommonHeaderFacetSettings {
}
exports.ObjectPageHeaderSectionChart = ObjectPageHeaderSectionChart;
class ObjectPageHeaderSectionDataPoint extends CommonHeaderFacetSettings {
}
exports.ObjectPageHeaderSectionDataPoint = ObjectPageHeaderSectionDataPoint;
class ObjectPageHeaderSectionContact extends CommonHeaderFacetSettings {
}
exports.ObjectPageHeaderSectionContact = ObjectPageHeaderSectionContact;
class ObjectPageHeaderSectionAddress extends CommonHeaderFacetSettings {
}
exports.ObjectPageHeaderSectionAddress = ObjectPageHeaderSectionAddress;
class HeaderSections {
}
exports.HeaderSections = HeaderSections;
