// @sap/ux-specification/dist/src/sync/v4/export/controls/ObjectPageFooterAction.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomFooterActionOP = exports.CustomFooterActionPositionOP = exports.FooterActionV4 = exports.ActionAfterExecutionConfigurationFooter = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const control_1 = require("../../../common/decoration/control");
const application_1 = require("../../application");
class ActionAfterExecutionConfigurationFooter {
}
exports.ActionAfterExecutionConfigurationFooter = ActionAfterExecutionConfigurationFooter;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getFooterActionsPath)(pageName, manifest, sectionId, columnKey) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfigurationFooter.prototype, "navigateToInstance", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getFooterActionsPath)(pageName, manifest, sectionId, columnKey) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfigurationFooter.prototype, "enableAutoScroll", void 0);
class FooterActionV4 {
}
exports.FooterActionV4 = FooterActionV4;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getFooterActionsPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], FooterActionV4.prototype, "afterExecution", void 0);
/**
 * Sync class for Position
 */
class CustomFooterActionPositionOP {
}
exports.CustomFooterActionPositionOP = CustomFooterActionPositionOP;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getFooterActionsPath)(pageName, manifest, sectionId, columnKey) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomFooterActionPositionOP.prototype, "anchor", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getFooterActionsPath)(pageName, manifest, sectionId, columnKey) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomFooterActionPositionOP.prototype, "placement", void 0);
class CustomFooterActionOP {
}
exports.CustomFooterActionOP = CustomFooterActionOP;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getFooterActionsPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomFooterActionOP.prototype, "text", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getFooterActionsPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomFooterActionOP.prototype, "position", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getFooterActionsPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomFooterActionOP.prototype, "press", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getFooterActionsPath
        },
        generate: control_1.addPatternForBindingChangeOfEnumOPV4
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomFooterActionOP.prototype, "visible", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getFooterActionsPath
        },
        generate: control_1.addPatternForBindingChangeOfEnumOPV4
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomFooterActionOP.prototype, "enabled", void 0);
