// @sap/ux-specification/dist/src/sync/v4/export/controls/ObjectPageForm.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPageForm = exports.ObjectPageFormActions = exports.FormAction = exports.ActionAfterExecutionConfigurationForm = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
class ActionAfterExecutionConfigurationForm {
}
exports.ActionAfterExecutionConfigurationForm = ActionAfterExecutionConfigurationForm;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getActionsPathForOP)(pageName, manifest, sectionId, columnKey) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfigurationForm.prototype, "navigateToInstance", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getActionsPathForOP)(pageName, manifest, sectionId, columnKey) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfigurationForm.prototype, "enableAutoScroll", void 0);
class FormAction {
}
exports.FormAction = FormAction;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], FormAction.prototype, "afterExecution", void 0);
class ObjectPageFormActions {
}
exports.ObjectPageFormActions = ObjectPageFormActions;
class ObjectPageForm {
}
exports.ObjectPageForm = ObjectPageForm;
