// @sap/ux-specification/dist/src/sync/v4/export/controls/ObjectPageToolBarAction.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomTableActionOP = exports.CustomActionPositionOP = exports.ObjectPageToolBarAction = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
const control_1 = require("../../../common/decoration/control");
class ObjectPageToolBarAction {
}
exports.ObjectPageToolBarAction = ObjectPageToolBarAction;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForOP
        }
    })
], ObjectPageToolBarAction.prototype, "enableOnSelect", void 0);
/**
 * Sync class for Position
 */
class CustomActionPositionOP {
}
exports.CustomActionPositionOP = CustomActionPositionOP;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getActionsPathForOP)(pageName, manifest, sectionId, columnKey) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomActionPositionOP.prototype, "anchor", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getActionsPathForOP)(pageName, manifest, sectionId, columnKey) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomActionPositionOP.prototype, "placement", void 0);
class CustomTableActionOP {
}
exports.CustomTableActionOP = CustomTableActionOP;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableActionOP.prototype, "text", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableActionOP.prototype, "position", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableActionOP.prototype, "press", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForOP
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableActionOP.prototype, "requiresSelection", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForOP
        },
        generate: control_1.addPatternForBindingChangeOfEnumOPV4
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableActionOP.prototype, "visible", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForOP
        },
        generate: control_1.addPatternForBindingChangeOfEnumOPV4
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableActionOP.prototype, "enabled", void 0);
