// @sap/ux-specification/dist/src/sync/v4/export/controls/ObjectPageCustomSection.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPageCustomSubSectionFragment = exports.ObjectPageCustomSectionFragment = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
class ObjectPageCustomSectionFragment {
}
exports.ObjectPageCustomSectionFragment = ObjectPageCustomSectionFragment;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId) => {
                return (0, application_1.getCustomSectionPath)(pageName, manifest, sectionId, 'position');
            },
            key: 'anchor'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "relatedFacet", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: (_manifest, sectionId) => {
                // Id is read from path's section - it is object's key
                return sectionId;
            },
            export: false
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "id", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId) => (0, application_1.getCustomSectionPath)(pageName, manifest, sectionId, 'position'),
            key: 'placement'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "relativePosition", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "title", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: (manifest) => {
                // Currently only XML supported, but idea is that we need to check 'manifest.type' and allow other types - maybe JS, JSON, etc.
                if (!manifest.type || manifest.type.startsWith(ux_specification_types_1.v4.CustomSectionViewTypesV4.XML)) {
                    return ux_specification_types_1.v4.CustomSectionViewTypesV4.XML;
                }
                return '';
            },
            export: (manifest, object) => {
                // Currently only 'XMLFragment', but in future we can add logic here
                let type = '';
                if ('type' in object) {
                    type += object.type;
                }
                if ('className' in object) {
                    if (object.className === ux_specification_types_1.SAPUI5_FRAGMENT_CLASS) {
                        type += 'Fragment';
                    }
                    else {
                        type += object.className;
                    }
                }
                manifest.type = type;
            }
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "type", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: (manifest) => {
                // Currently only XML Fragment supported, but idea is that we need to check 'value.type'
                // f.e. type === 'XMLFragment' ? 'sap.ui.core.Fragment' : ...
                if (!manifest.type || manifest.type.startsWith(ux_specification_types_1.v4.CustomSectionViewTypesV4.XML)) {
                    if (!manifest.type || manifest.type === 'XMLFragment') {
                        return ux_specification_types_1.SAPUI5_FRAGMENT_CLASS;
                    }
                    else {
                        return manifest.type.slice(ux_specification_types_1.v4.CustomSectionViewTypesV4.XML.length);
                    }
                }
                // Fallback - if there is unsupported value
                return manifest.type;
            },
            export: false
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "className", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: (manifestSection) => {
                // During import, we need to consider both `name` and `template` to correctly fill our config and to be upgrade safe
                return manifestSection.name || manifestSection.template;
            },
            key: 'template' // key: 'name' deprecated since '1.86'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSectionFragment.prototype, "fragmentName", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSectionPath,
            import: false,
            export: false
        },
        xml: (config) => {
            return config['fragmentName'];
        }
    }),
    (0, decorators_1.validity)({
        since: '1.96.0'
    })
], ObjectPageCustomSectionFragment.prototype, "controls", void 0);
class ObjectPageCustomSubSectionFragment extends ObjectPageCustomSectionFragment {
}
exports.ObjectPageCustomSubSectionFragment = ObjectPageCustomSubSectionFragment;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (...args) => {
                const resolvedPath = application_1.getCustomSubSectionPath.apply({}, args);
                return `${resolvedPath}/position`;
            },
            key: 'anchor'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "relatedFacet", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (...args) => {
                const resolvedPath = application_1.getCustomSubSectionPath.apply({}, args);
                return `${resolvedPath}/position`;
            },
            key: 'placement'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "relativePosition", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSubSectionPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "title", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSubSectionPath,
            import: (manifestSection) => {
                // During import, we need to consider both `name` and `template` to correctly fill our config and to be upgrade safe
                return manifestSection.name || manifestSection.template;
            },
            key: 'template' // key: 'name' deprecated since '1.86'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "fragmentName", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getCustomSubSectionPath,
            import: false,
            export: false
        },
        xml: (config) => {
            return config['fragmentName'];
        }
    }),
    (0, decorators_1.validity)({
        since: '1.96.0'
    })
], ObjectPageCustomSubSectionFragment.prototype, "controls", void 0);
