// @sap/ux-specification/dist/src/sync/v4/export/controls/ObjectPageHeaderAction.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomHeaderActionOP = exports.CustomHeaderActionPositionOP = exports.ObjectPageHeaderAction = exports.ActionAfterExecutionConfigurationObjectPageHeader = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const control_1 = require("../../../common/decoration/control");
const application_1 = require("../../application");
const HeaderAction_1 = require("./HeaderAction");
class ActionAfterExecutionConfigurationObjectPageHeader {
}
exports.ActionAfterExecutionConfigurationObjectPageHeader = ActionAfterExecutionConfigurationObjectPageHeader;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getHeaderActionsPath)(pageName, manifest, sectionId, columnKey) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfigurationObjectPageHeader.prototype, "navigateToInstance", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey) => (0, application_1.getHeaderActionsPath)(pageName, manifest, sectionId, columnKey) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfigurationObjectPageHeader.prototype, "enableAutoScroll", void 0);
class ObjectPageHeaderAction {
}
exports.ObjectPageHeaderAction = ObjectPageHeaderAction;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getHeaderActionsPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ObjectPageHeaderAction.prototype, "afterExecution", void 0);
/**
 * Sync class for Position
 */
class CustomHeaderActionPositionOP extends HeaderAction_1.CustomHeaderActionPosition {
}
exports.CustomHeaderActionPositionOP = CustomHeaderActionPositionOP;
class CustomHeaderActionOP {
}
exports.CustomHeaderActionOP = CustomHeaderActionOP;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getHeaderActionsPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomHeaderActionOP.prototype, "text", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getHeaderActionsPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomHeaderActionOP.prototype, "position", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getHeaderActionsPath
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomHeaderActionOP.prototype, "press", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getHeaderActionsPath
        },
        generate: control_1.addPatternForBindingChangeOfEnumOPV4
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomHeaderActionOP.prototype, "visible", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getHeaderActionsPath
        },
        generate: control_1.addPatternForBindingChangeOfEnumOPV4
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomHeaderActionOP.prototype, "enabled", void 0);
