// @sap/ux-specification/dist/src/sync/v4/export/controls/TableColumn.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewTableCustomColumn = exports.TableCustomColumn = exports.ViewPosition = exports.Position = exports.ViewTableColumnAction = exports.TableColumnAction = exports.ViewTableColumn = exports.TableColumn = exports.ActionAfterExecutionConfiguration = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
class ActionAfterExecutionConfiguration {
}
exports.ActionAfterExecutionConfiguration = ActionAfterExecutionConfiguration;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getColumnsPathForLR)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfiguration.prototype, "navigateToInstance", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getColumnsPathForLR)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfiguration.prototype, "enableAutoScroll", void 0);
/************************************
 * Sync class for TableColumn
 */
class TableColumn {
}
exports.TableColumn = TableColumn;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableColumn.prototype, "width", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.85.0'
    })
], TableColumn.prototype, "availability", void 0);
/************************************
 * Sync class for ViewTableColumn
 */
class ViewTableColumn {
}
exports.ViewTableColumn = ViewTableColumn;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableColumn.prototype, "width", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.85.0'
    })
], ViewTableColumn.prototype, "availability", void 0);
class TableColumnAction extends TableColumn {
}
exports.TableColumnAction = TableColumnAction;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableColumnAction.prototype, "afterExecution", void 0);
class ViewTableColumnAction extends ViewTableColumn {
}
exports.ViewTableColumnAction = ViewTableColumnAction;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableColumnAction.prototype, "afterExecution", void 0);
/**
 * Sync class for Position
 */
class Position {
}
exports.Position = Position;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getColumnsPathForLR)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], Position.prototype, "anchor", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getColumnsPathForLR)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], Position.prototype, "placement", void 0);
/**
 * Sync class for ViewPosition
 */
class ViewPosition {
}
exports.ViewPosition = ViewPosition;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getColumnsPathForView)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewPosition.prototype, "anchor", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getColumnsPathForView)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewPosition.prototype, "placement", void 0);
/**
 * Sync class for TableCustomColumn
 */
class TableCustomColumn {
}
exports.TableCustomColumn = TableCustomColumn;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableCustomColumn.prototype, "position", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableCustomColumn.prototype, "header", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableCustomColumn.prototype, "width", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableCustomColumn.prototype, "template", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.86.0'
    })
], TableCustomColumn.prototype, "horizontalAlign", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], TableCustomColumn.prototype, "availability", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.85.0'
    })
], TableCustomColumn.prototype, "properties", void 0);
/**
 * Sync class for ViewTableCustomColumn
 */
class ViewTableCustomColumn {
}
exports.ViewTableCustomColumn = ViewTableCustomColumn;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomColumn.prototype, "position", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomColumn.prototype, "header", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomColumn.prototype, "width", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomColumn.prototype, "template", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.86.0'
    })
], ViewTableCustomColumn.prototype, "horizontalAlign", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomColumn.prototype, "availability", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getColumnsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.85.0'
    })
], ViewTableCustomColumn.prototype, "properties", void 0);
