// @sap/ux-specification/dist/src/sync/v4/export/controls/ToolBarAction.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewTableCustomAction = exports.ViewCustomActionPosition = exports.CustomTableAction = exports.CustomActionPosition = exports.ViewToolBarAction = exports.ToolBarAction = exports.ActionAfterExecutionConfigurationToolBarLR = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
const control_1 = require("../../../common/decoration/control");
class ActionAfterExecutionConfigurationToolBarLR {
}
exports.ActionAfterExecutionConfigurationToolBarLR = ActionAfterExecutionConfigurationToolBarLR;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getActionsPathForLR)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfigurationToolBarLR.prototype, "navigateToInstance", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getActionsPathForLR)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/afterExecution'
        }
    })
], ActionAfterExecutionConfigurationToolBarLR.prototype, "enableAutoScroll", void 0);
class ToolBarAction {
}
exports.ToolBarAction = ToolBarAction;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForLR
        }
    })
], ToolBarAction.prototype, "enableOnSelect", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForLR
        }
    })
], ToolBarAction.prototype, "afterExecution", void 0);
class ViewToolBarAction {
}
exports.ViewToolBarAction = ViewToolBarAction;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForView
        }
    })
], ViewToolBarAction.prototype, "enableOnSelect", void 0);
/**
 * Sync class for Position
 */
class CustomActionPosition {
}
exports.CustomActionPosition = CustomActionPosition;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getActionsPathForLR)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomActionPosition.prototype, "anchor", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getActionsPathForLR)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomActionPosition.prototype, "placement", void 0);
class CustomTableAction {
}
exports.CustomTableAction = CustomTableAction;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableAction.prototype, "text", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableAction.prototype, "position", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableAction.prototype, "press", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForLR
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableAction.prototype, "requiresSelection", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForLR
        },
        generate: control_1.addPatternForBindingChangeOfEnumLR
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableAction.prototype, "visible", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForLR
        },
        generate: control_1.addPatternForBindingChangeOfEnumLR
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], CustomTableAction.prototype, "enabled", void 0);
/**
 * Sync class for Position
 */
class ViewCustomActionPosition {
}
exports.ViewCustomActionPosition = ViewCustomActionPosition;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getActionsPathForView)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewCustomActionPosition.prototype, "anchor", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName, manifest, sectionId, columnKey, targetAnnotation) => (0, application_1.getActionsPathForView)(pageName, manifest, sectionId, columnKey, targetAnnotation) + '/position'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewCustomActionPosition.prototype, "placement", void 0);
class ViewTableCustomAction {
}
exports.ViewTableCustomAction = ViewTableCustomAction;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomAction.prototype, "text", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomAction.prototype, "position", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomAction.prototype, "press", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForView
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomAction.prototype, "requiresSelection", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForView
        },
        generate: control_1.addPatternForBindingChangeOfEnumLR
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomAction.prototype, "visible", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: application_1.getActionsPathForView
        },
        generate: control_1.addPatternForBindingChangeOfEnumLR
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ViewTableCustomAction.prototype, "enabled", void 0);
