// @sap-ux/system-access/dist/base/prompts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.inquirer = exports.questions = void 0;
const fs_1 = require("fs");
const username = {
    type: 'text',
    name: 'username',
    message: 'Username:'
};
const password = {
    type: 'password',
    name: 'password',
    message: 'Password:'
};
const serviceKeysPath = {
    type: 'text',
    name: 'path',
    message: 'Please provide the service keys as file:',
    validate: (input) => (0, fs_1.existsSync)(input)
};
const storeCredentials = {
    type: 'confirm',
    name: 'store',
    message: 'Do you want to store your credentials in the secure storage?',
    initial: true
};
const systemName = {
    type: 'text',
    name: 'name',
    message: 'System name:',
    validate: (input) => !!input
};
/**
 * Export map of questions for usage with the prompts modules
 */
exports.questions = {
    username,
    password,
    serviceKeysPath,
    storeCredentials,
    systemName
};
/**
 * Generate a map of questions for usages with inquirer (e.g. in yeoman)
 */
exports.inquirer = {};
for (const key in exports.questions) {
    const question = exports.questions[key];
    exports.inquirer[key] = {
        type: question.type === 'text' ? 'input' : question.type,
        name: question.name,
        message: question.message,
        validate: question.validate,
        default: question.initial
    };
}
