// @sap-ux/store/dist/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFioriToolsDirectory = exports.FioriToolsSettings = exports.errorString = exports.errorInstance = exports.pick = void 0;
const os_1 = require("os");
const path_1 = __importDefault(require("path"));
/** Pick the properties listed and return a new object with a shallow-copy */
const pick = (target, ...props) => {
    return ((target &&
        (props === null || props === void 0 ? void 0 : props.length) > 0 &&
        props.reduce((o, k) => {
            o[k] = target[k];
            return o;
        }, {})) ||
        undefined);
};
exports.pick = pick;
/** Given an `Error` or any other object thrown, returns an `Error` instance */
function errorInstance(e) {
    if (e instanceof Error) {
        return e;
    }
    else {
        return new Error(String(e));
    }
}
exports.errorInstance = errorInstance;
/** If input in an instance of `Error` return the message property,
 *  otherwise convert the input to its string representation
 */
function errorString(e) {
    return e instanceof Error ? e.message : String(e);
}
exports.errorString = errorString;
var FioriToolsSettings;
(function (FioriToolsSettings) {
    FioriToolsSettings["dir"] = ".fioritools";
})(FioriToolsSettings || (exports.FioriToolsSettings = FioriToolsSettings = {}));
const getFioriToolsDirectory = () => {
    return path_1.default.join((0, os_1.homedir)(), FioriToolsSettings.dir);
};
exports.getFioriToolsDirectory = getFioriToolsDirectory;
__exportStar(require("./app-studio"), exports);
