// @sap-ux/store/dist/data-provider/api-hub.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiHubSettingsProvider = void 0;
const hybrid_1 = require("../data-access/hybrid");
const constants_1 = require("./constants");
const api_hub_1 = require("../entities/api-hub");
const ApiHubSettingsProvider = class {
    constructor(logger, options = {}) {
        this.entityName = constants_1.Entities.ApiHub;
        this.logger = logger;
        this.dataAccessor = (0, hybrid_1.getHybridStore)(this.logger, options);
    }
    read(key) {
        return this.dataAccessor.read({ entityName: this.entityName, id: key.getId() });
    }
    write(entity) {
        return this.dataAccessor.write({
            entityName: this.entityName,
            id: api_hub_1.ApiHubSettingsKey.SINGLETON,
            entity
        });
    }
    delete(_entity) {
        return this.dataAccessor.del({
            entityName: this.entityName,
            id: api_hub_1.ApiHubSettingsKey.SINGLETON
        });
    }
    getAll() {
        return this.dataAccessor.getAll({ entityName: this.entityName });
    }
};
exports.ApiHubSettingsProvider = ApiHubSettingsProvider;
