// @sap/cf-tools/out/src/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.eServiceTypes = exports.eOrderDirection = exports.eOperation = exports.eFilters = exports.CF_CMD_EXIT_CODE = exports.DEFAULT_TARGET = exports.CF_PAGE_SIZE = exports.NEW_LINE = exports.OK = void 0;
exports.OK = "OK";
exports.NEW_LINE = "\n";
exports.CF_PAGE_SIZE = 99 * 3;
exports.DEFAULT_TARGET = "Default (no targets)";
var CF_CMD_EXIT_CODE;
(function (CF_CMD_EXIT_CODE) {
    CF_CMD_EXIT_CODE[CF_CMD_EXIT_CODE["OK"] = 0] = "OK";
    CF_CMD_EXIT_CODE[CF_CMD_EXIT_CODE["ERROR"] = -1] = "ERROR";
    CF_CMD_EXIT_CODE[CF_CMD_EXIT_CODE["CANCEL_REQ"] = -2] = "CANCEL_REQ";
    CF_CMD_EXIT_CODE[CF_CMD_EXIT_CODE["CANCELED"] = -3] = "CANCELED";
})(CF_CMD_EXIT_CODE = exports.CF_CMD_EXIT_CODE || (exports.CF_CMD_EXIT_CODE = {}));
var eFilters;
(function (eFilters) {
    eFilters["type"] = "type";
    eFilters["names"] = "names";
    eFilters["guids"] = "guids";
    eFilters["app_guids"] = "app_guids";
    eFilters["app_names"] = "app_names";
    eFilters["space_guids"] = "space_guids";
    eFilters["available"] = "available";
    eFilters["broker_catalog_ids"] = "broker_catalog_ids";
    eFilters["service_broker_guids"] = "service_broker_guids";
    eFilters["service_broker_names"] = "service_broker_names";
    eFilters["service_plan_guids"] = "service_plan_guids";
    eFilters["organization_guids"] = "organization_guids";
    eFilters["service_plan_names"] = "service_plan_names";
    eFilters["service_plan"] = "service_plan";
    eFilters["service_instance_guids"] = "service_instance_guids";
    eFilters["service_instance_names"] = "service_instance_names";
    eFilters["service_offering_guids"] = "service_offering_guids";
    eFilters["service_offering_names"] = "service_offering_names";
    eFilters["label_selector"] = "label_selector";
    eFilters["page"] = "page";
    eFilters["per_page"] = "per_page";
    eFilters["oder_by"] = "order_by";
    eFilters["created_ats"] = "created_ats";
    eFilters["updated_ats"] = "updated_ats";
    eFilters["status"] = "status";
    eFilters["include"] = "include";
})(eFilters = exports.eFilters || (exports.eFilters = {}));
var eOperation;
(function (eOperation) {
    eOperation["gte"] = "gte";
    eOperation["lte"] = "lte";
    eOperation["lt"] = "lt";
    eOperation["gt"] = "gt";
    eOperation["not"] = "not";
    eOperation["fields"] = "fields";
})(eOperation = exports.eOperation || (exports.eOperation = {}));
var eOrderDirection;
(function (eOrderDirection) {
    eOrderDirection[eOrderDirection["asc"] = 0] = "asc";
    eOrderDirection[eOrderDirection["desc"] = 1] = "desc";
})(eOrderDirection = exports.eOrderDirection || (exports.eOrderDirection = {}));
var eServiceTypes;
(function (eServiceTypes) {
    eServiceTypes["managed"] = "managed";
    eServiceTypes["user_provided"] = "user-provided";
})(eServiceTypes = exports.eServiceTypes || (exports.eServiceTypes = {}));
