// @sap/ux-specification/dist/src/sync/v2/export/pages/AnalyticalListPage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnalyticalListPage = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decorators_1 = require("../../../common/decoration/decorators");
const decoration_1 = require("../../../common/decoration");
let AnalyticalListPage = class AnalyticalListPage {
};
exports.AnalyticalListPage = AnalyticalListPage;
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: decoration_1.buildControlIdForPage,
            controlType: () => ux_specification_types_1.ControlType.DynamicPage,
            createFlexChange: decoration_1.exportToFlexChange
        }
    })
], AnalyticalListPage.prototype, "fitContent", void 0);
__decorate([
    (0, decorators_1.childId)('table')
], AnalyticalListPage.prototype, "table", void 0);
__decorate([
    (0, decorators_1.childId)('template::SmartFilterBar')
], AnalyticalListPage.prototype, "filterBar", void 0);
__decorate([
    (0, decorators_1.childId)('analyticalListPageChart')
], AnalyticalListPage.prototype, "chart", void 0);
__decorate([
    (0, decorators_1.childId)('analyticalListFooter')
], AnalyticalListPage.prototype, "footer", void 0);
exports.AnalyticalListPage = AnalyticalListPage = __decorate([
    (0, decorators_1.pageLayoutInformation)({
        id: 'sap.suite.ui.generic.template.AnalyticalListPage.view.AnalyticalListPage',
        pageId: ux_specification_types_1.v2.FE_TEMPLATE_V2_ALP
    })
], AnalyticalListPage);
