// @sap/ux-specification/dist/src/sync/v2/export/pages/ObjectPage.js
"use strict";
/**
 * The module is mapped to specification/v2/pages/ObjectPageConfig.ts
 * This comment is maintained as this module is named as ObjectPage.ts and the corresponding module is named as ObjectPageConfig.ts
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPage = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
const common_1 = require("../../../common");
let ObjectPage = class ObjectPage {
};
exports.ObjectPage = ObjectPage;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppSettingOBJPathV2)(pageKeys)
        }
    }),
    (0, decorators_1.validity)({
        since: '1.52.0'
    })
], ObjectPage.prototype, "showConfirmationOnDraftActivate", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppSettingOBJPathV2)(pageKeys)
        }
    }),
    (0, decorators_1.validity)({
        since: '1.52.1'
    })
], ObjectPage.prototype, "hideChevronForUnauthorizedExtNav", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppSettingOBJPathV2)(pageKeys)
        }
    }),
    (0, decorators_1.validity)({
        since: '1.48.0'
    })
], ObjectPage.prototype, "allowDeepLinking", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppSettingOBJPathV2)(pageKeys)
        }
    }),
    (0, decorators_1.validity)({
        since: '1.52.0'
    })
], ObjectPage.prototype, "navToListOnSave", void 0);
__decorate([
    (0, decorators_1.childId)('header')
], ObjectPage.prototype, "header", void 0);
__decorate([
    (0, decorators_1.childId)('sections')
], ObjectPage.prototype, "sections", void 0);
__decorate([
    (0, decorators_1.childId)('layout')
], ObjectPage.prototype, "layout", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppRootSettingOBJPath)(pageKeys)
        }
    }),
    (0, decorators_1.validity)({
        since: '1.77.0'
    })
], ObjectPage.prototype, "defaultLayoutTypeIfExternalNavigation", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppSettingOBJPathV2)(pageKeys),
            import: (value) => {
                if ('chartSettings' in value) {
                    return value['chartSettings'];
                }
            },
            key: 'chartSettings'
        }
    }),
    (0, decorators_1.validity)({
        since: '1.78.0'
    })
], ObjectPage.prototype, "globalChartSettings", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppSettingOBJPathV2)(pageKeys),
            import: (value) => {
                if ('tableSettings' in value) {
                    return value['tableSettings'];
                }
            },
            key: 'tableSettings',
            delete: common_1.doNotDelete
        }
    }),
    (0, decorators_1.validity)({
        since: '1.71.0'
    })
], ObjectPage.prototype, "globalTableSettings", void 0);
exports.ObjectPage = ObjectPage = __decorate([
    (0, decorators_1.pageLayoutInformation)({
        id: ux_specification_types_1.v2.SAPUI5_VIEW_EXTENSION_OBJECT_PAGE,
        pageId: ux_specification_types_1.v2.FE_TEMPLATE_V2_OBJECT_PAGE
    })
], ObjectPage);
