// @sap-ux/store/dist/entities/system-migration-status.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SystemMigrationStatusKey = exports.SystemMigrationStatus = void 0;
const decorators_1 = require("../decorators");
/** This represents the migration status of the systems from pure secure store
 * the hybrid filesystem/secure store
 */
class SystemMigrationStatus {
    constructor({ authTypeMigrated, migrationDone, migrationLogs }) {
        this.migrationDone = migrationDone;
        this.authTypeMigrated = authTypeMigrated;
        this.migrationLogs = migrationLogs;
    }
}
exports.SystemMigrationStatus = SystemMigrationStatus;
__decorate([
    decorators_1.serializable,
    __metadata("design:type", Boolean)
], SystemMigrationStatus.prototype, "migrationDone", void 0);
__decorate([
    decorators_1.serializable,
    __metadata("design:type", Boolean)
], SystemMigrationStatus.prototype, "authTypeMigrated", void 0);
__decorate([
    decorators_1.serializable,
    __metadata("design:type", Array)
], SystemMigrationStatus.prototype, "migrationLogs", void 0);
class SystemMigrationStatusKey {
    getId() {
        return 'systemMigration'; // A singleton
    }
}
exports.SystemMigrationStatusKey = SystemMigrationStatusKey;
