// @sap/ux-specification/dist/src/sync/v2/generate/overviewPage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateOverviewPageSchemaV2 = exports.addModelsEnum = exports.getCardDefinitionNames = void 0;
const common_1 = require("../../common");
/**
 * Method returns array of card definition names.
 * @param genericSchema - Generic schema of Overview Page.
 *
 * @returns Array of definition names
 */
function getCardDefinitionNames(genericSchema) {
    // Apply enum to all card properties
    const properties = genericSchema['properties'];
    const cardProperties = (properties['cards'] &&
        properties['cards']['additionalProperties'] &&
        properties['cards']['additionalProperties']['anyOf']) ||
        [];
    return cardProperties.map((cardProperty) => {
        // Resolve ref
        const refParts = cardProperty['$ref'].split('/');
        return refParts[refParts.length - 1];
    });
}
exports.getCardDefinitionNames = getCardDefinitionNames;
/**
 * Defines and adds enum entries for 'model' property of Cards.
 * @param appSchema - the app specific schema that shall get enhanced
 * @param manifest - manifest.json of the app
 */
function addModelsEnum(appSchema, manifest) {
    const dataSources = (manifest && manifest['sap.app'] && manifest['sap.app'].dataSources) || {};
    const serviceKeys = Object.keys(dataSources);
    const models = (manifest && manifest['sap.ui5'] && manifest['sap.ui5'].models) || {};
    let enumValues = [];
    enumValues = enumValues.concat(Object.keys(models).filter((model) => serviceKeys.includes(models[model].dataSource)));
    // Define enum
    const definition = appSchema.definitions;
    definition['CardModel'] = {
        type: 'string'
    };
    (0, common_1.addEnumToSchema)(enumValues, definition['CardModel']);
    // Apply enum to all card properties
    const cards = getCardDefinitionNames(appSchema);
    for (const card of cards) {
        const property = definition[card]['properties']['model'];
        delete property.type;
        property['$ref'] = common_1.DEFINITION_LINK_PREFIX + 'CardModel';
    }
}
exports.addModelsEnum = addModelsEnum;
/**
 * Generates an app specific schema out of the generic schema.
 * Generic types are replaced by information from the app manifest.
 *
 * @param genericSchema - generic JSON schema of an object page
 * @param manifest - manifest.json of the app
 *
 * @returns the app specific JSON schema
 */
function generateOverviewPageSchemaV2(genericSchema, manifest) {
    const appSchema = JSON.parse(JSON.stringify(genericSchema));
    // schema validation for card name
    appSchema['properties']['cards']['propertyNames'] = {
        pattern: '^[a-zA-Z0-9_\\.\\-]+$'
    };
    // Enhance 'model' property with enum entries
    addModelsEnum(appSchema, manifest);
    return appSchema;
}
exports.generateOverviewPageSchemaV2 = generateOverviewPageSchemaV2;
