// @sap-ux/store/dist/i18n.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.text = exports.initI18n = void 0;
const i18next_1 = __importDefault(require("i18next"));
const i18next_fs_backend_1 = __importDefault(require("i18next-fs-backend"));
const path_1 = __importDefault(require("path"));
const i18nInstance = i18next_1.default.createInstance();
function initI18n() {
    return __awaiter(this, void 0, void 0, function* () {
        yield i18nInstance.use(i18next_fs_backend_1.default).init({
            initImmediate: false,
            fallbackLng: 'en',
            fallbackNS: 'default',
            interpolation: { escapeValue: false },
            ns: 'ux-store',
            defaultNS: 'ux-store',
            backend: {
                loadPath: path_1.default.join(__dirname, './translations/{{lng}}.{{ns}}.json')
            }
        });
    });
}
exports.initI18n = initI18n;
function text(key, options) {
    return i18nInstance.t(key, options);
}
exports.text = text;
