// @sap/ux-specification/dist/src/sync/v2/export/controls/Field.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Field = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decoration_1 = require("../../../common/decoration");
const common_1 = require("../../../common");
/**
 * Create stable ID for Object Page Form Field
 *
 * @param {string} baseId Base Id of the control
 * @param {string[]} idBreadcrumbs
 * @param {string[]} breadcrumbs
 * @param {ControlTypeFunction} controlType
 * @param {string} facetID title from the app schema (comprising for instance the facet ID)
 * @returns {string} stable ID of Object Page Form Action button
 */
const getSmartFieldControlId = (baseId, _idBreadcrumbs, breadcrumbs, _controlType, facetId) => {
    const convertedId = (0, common_1.convertSectionId)(breadcrumbs[1], facetId);
    const fieldCrumb = breadcrumbs.slice(-1)[0];
    const breadcrumbsTrail = [convertedId, fieldCrumb, 'GroupElement'];
    return (0, decoration_1.buildControlIdFromParent)(baseId, breadcrumbsTrail);
};
class Field {
}
exports.Field = Field;
__decorate([
    (0, decoration_1.syncRule)({
        flex: {
            controlId: getSmartFieldControlId,
            controlType: () => ux_specification_types_1.ControlType.GroupElement,
            createFlexChange: decoration_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], Field.prototype, "visible", void 0);
