// @sap/ux-specification/dist/src/sync/v2/export/controls/FormAction.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldForAction = exports.FormAction = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decoration_1 = require("../../../common/decoration");
const common_1 = require("../../../common");
/**
 * Create stable ID for Object Page Form Action button
 *
 * @param {string} baseId Base Id of the control
 * @param {string[]} idBreadcrumbs
 * @param {string[]} breadcrumbs
 * @param {ControlTypeFunction} controlType
 * @param {string} facetID title from the app schema (comprising for instance the facet ID)
 * @returns {string} stable ID of Object Page Form Action button
 */
const getFormActionButtonId = (baseId, idBreadcrumbs, breadcrumbs, controlType, facetID) => {
    if (controlType() === ux_specification_types_1.ControlType.Button) {
        const entity = breadcrumbs[breadcrumbs.length - 1];
        const section = (0, decoration_1.convertSectionId)(breadcrumbs[1], facetID);
        return `${baseId}action::${entity}::${section}::FormAction`;
    }
    return (0, decoration_1.buildControlIdFromParent)(baseId, idBreadcrumbs);
};
const flexRuleForFormAction = {
    controlId: getFormActionButtonId,
    controlType: () => ux_specification_types_1.ControlType.Button,
    createFlexChange: decoration_1.exportToFlexChange
};
class FormAction {
}
exports.FormAction = FormAction;
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFormAction,
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], FormAction.prototype, "visible", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFormAction
    })
], FormAction.prototype, "tooltip", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFormAction
    })
], FormAction.prototype, "icon", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFormAction
    })
], FormAction.prototype, "activeIcon", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFormAction
    })
], FormAction.prototype, "type", void 0);
class FieldForAction {
}
exports.FieldForAction = FieldForAction;
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFormAction,
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], FieldForAction.prototype, "visible", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFormAction
    })
], FieldForAction.prototype, "tooltip", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFormAction
    })
], FieldForAction.prototype, "icon", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFormAction
    })
], FieldForAction.prototype, "activeIcon", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFormAction
    }),
    (0, decoration_1.validity)({
        since: '1.73.0'
    })
], FieldForAction.prototype, "type", void 0);
