// @sap/ux-specification/dist/src/sync/v2/export/controls/ObjectPageFooterAction.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPageFooterAction = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decoration_1 = require("../../../common/decoration");
const common_1 = require("../../../common");
/**
 * Create stable ID for Object Page Footer Action button
 *
 * @param {string} baseId Base Id of the control
 * @param {string[]} idBreadcrumbs
 * @param {string[]} breadcrumbs
 * @param {ControlTypeFunction} controlType
 * @returns {string} stable ID of Object Page Footer Action button
 */
const getFooterActionButtonId = (baseId, idBreadcrumbs, breadcrumbs, controlType) => {
    if (controlType() === ux_specification_types_1.ControlType.Button) {
        const { actionId } = (0, common_1.getActionBreadcrumbParts)(breadcrumbs);
        return `${baseId}action::${(0, common_1.convertActionKeyToStableId)(actionId)}::Determining`;
    }
    return (0, decoration_1.buildControlIdFromParent)(baseId, idBreadcrumbs);
};
const flexRuleForFooterAction = {
    controlId: getFooterActionButtonId,
    controlType: () => ux_specification_types_1.ControlType.Button,
    createFlexChange: decoration_1.exportToFlexChange
};
class ObjectPageFooterAction {
}
exports.ObjectPageFooterAction = ObjectPageFooterAction;
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFooterAction,
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageFooterAction.prototype, "visible", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFooterAction
    })
], ObjectPageFooterAction.prototype, "tooltip", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFooterAction
    })
], ObjectPageFooterAction.prototype, "icon", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFooterAction
    })
], ObjectPageFooterAction.prototype, "activeIcon", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForFooterAction
    })
], ObjectPageFooterAction.prototype, "type", void 0);
