// @sap/ux-specification/dist/src/sync/v2/export/controls/ObjectPageToolBarAction.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPageToolBarAction = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decoration_1 = require("../../../common/decoration");
const common_1 = require("../../../common");
/**
 * Create stable ID for Object Page Toolbar Action button
 *
 * @param {string} baseId Base Id of the control
 * @param {string[]} _idBreadcrumbs
 * @param {string[]} breadcrumbs
 * @param {ControlTypeFunction} controlType
 * @param {string} facetID title from the app schema (comprising for instance the facet ID)
 * @returns {string} stable ID of Object Page Toolbar Action button
 */
const getActionButtonId = (baseId, _idBreadcrumbs, breadcrumbs, _controlType, facetID) => {
    const { actionId } = (0, common_1.getActionBreadcrumbParts)(breadcrumbs);
    const section = (0, common_1.convertSectionId)(breadcrumbs[1], facetID);
    return `${baseId}${section}::action::${(0, common_1.convertActionKeyToStableId)(actionId).replace(':2f', '::')}`;
};
const flexRuleForToolBarActionOP = {
    controlId: getActionButtonId,
    controlType: () => ux_specification_types_1.ControlType.Button,
    createFlexChange: decoration_1.exportToFlexChange
};
class ObjectPageToolBarAction {
}
exports.ObjectPageToolBarAction = ObjectPageToolBarAction;
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForToolBarActionOP,
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageToolBarAction.prototype, "visible", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForToolBarActionOP
    })
], ObjectPageToolBarAction.prototype, "tooltip", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForToolBarActionOP
    })
], ObjectPageToolBarAction.prototype, "icon", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForToolBarActionOP
    })
], ObjectPageToolBarAction.prototype, "activeIcon", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: flexRuleForToolBarActionOP
    })
], ObjectPageToolBarAction.prototype, "type", void 0);
