// @sap/ux-specification/dist/src/sync/v2/export/controls/ObjectPageHeader.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPageHeader = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decoration_1 = require("../../../common/decoration");
const application_1 = require("../../application");
const common_1 = require("../../../common");
const getAvatarControlId = (baseId, idBreadcrumbs) => {
    const idBreadcrumbsCopy = idBreadcrumbs.slice(0);
    idBreadcrumbsCopy[idBreadcrumbs.length - 1] = 'objectImage';
    return (0, decoration_1.buildControlIdFromParent)(baseId, idBreadcrumbsCopy);
};
class ObjectPageHeader {
}
exports.ObjectPageHeader = ObjectPageHeader;
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppSettingOBJPathV2)(pageKeys)
        }
    }),
    (0, decoration_1.validity)({
        since: '1.46.0'
    })
], ObjectPageHeader.prototype, "showRelatedApps", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppSettingOBJPathV2)(pageKeys)
        }
    }),
    (0, decoration_1.validity)({
        since: '1.46.0'
    })
], ObjectPageHeader.prototype, "editableHeaderContent", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppSettingOBJPathV2)(pageKeys)
        }
    }),
    (0, decoration_1.validity)({
        since: '1.48.0'
    })
], ObjectPageHeader.prototype, "simpleHeaderFacets", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: (pageKeys) => (0, application_1.getAppSettingOBJPathV2)(pageKeys)
        }
    }),
    (0, decoration_1.validity)({
        since: '1.48.0'
    })
], ObjectPageHeader.prototype, "showHeaderAsCarouselOnDesktop", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        flex: {
            controlId: getAvatarControlId,
            controlType: () => ux_specification_types_1.ControlType.Avatar,
            createFlexChange: decoration_1.exportToFlexChange
        },
        generate: common_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageHeader.prototype, "displayShape", void 0);
