// @sap/ux-specification/dist/src/sync/v2/export/controls/ObjectPageLayout.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPageLayout = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decorators_1 = require("../../../common/decoration/decorators");
const decoration_1 = require("../../../common/decoration");
const getControlId = (baseId, idBreadcrumbs) => {
    const idBreadcrumbsCopy = idBreadcrumbs.slice(0);
    idBreadcrumbsCopy[idBreadcrumbs.length - 1] = 'objectPage';
    return (0, decoration_1.buildControlIdFromParent)(baseId, idBreadcrumbsCopy);
};
class ObjectPageLayout {
}
exports.ObjectPageLayout = ObjectPageLayout;
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: getControlId,
            controlType: () => ux_specification_types_1.ControlType.ObjectPageLayout,
            createFlexChange: decoration_1.exportToFlexChange
        },
        generate: decoration_1.addPatternForBindingChangeOfEnumOP
    })
], ObjectPageLayout.prototype, "showHeaderContent", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: getControlId,
            controlType: () => ux_specification_types_1.ControlType.ObjectPageLayout,
            createFlexChange: decoration_1.exportToFlexChange
        }
    })
], ObjectPageLayout.prototype, "showAnchorBar", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: getControlId,
            controlType: () => ux_specification_types_1.ControlType.ObjectPageLayout,
            createFlexChange: decoration_1.exportToFlexChange
        }
    })
], ObjectPageLayout.prototype, "useIconTabBar", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        flex: {
            controlId: getControlId,
            controlType: () => ux_specification_types_1.ControlType.ObjectPageLayout,
            createFlexChange: decoration_1.exportToFlexChange
        }
    }),
    (0, decorators_1.validity)({
        since: '1.34.0'
    })
], ObjectPageLayout.prototype, "alwaysShowContentHeader", void 0);
