// @sap/ux-specification/dist/src/sync/v4/export/pages/ListReport.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListReport = void 0;
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
/**
 * sync rule deleting all views if the chart view gets deleted
 * @param manifestSection - manifest section of the given view
 */
function deleteAllALPViews(manifestSection) {
    if (manifestSection['paths']?.[0]?.primary || manifestSection['paths']?.[0]?.secondary) {
        Object.keys(manifestSection).forEach((key) => delete manifestSection[key]);
    }
}
/**
 * sync rule for preventing the export of a property
 */
function doNotExport() {
    return;
}
/**
 * sync rule for preventing the import of a property
 */
function doNotImport() {
    return;
}
/**
 * Sync class for import and export of V4 list report
 */
let ListReport = class ListReport {
};
exports.ListReport = ListReport;
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName) + '/views',
            import: doNotImport,
            export: doNotExport,
            delete: deleteAllALPViews
        }
    })
], ListReport.prototype, "chart", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName)
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ListReport.prototype, "variantManagement", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName) + '/views/paths/[]/0'
        }
    })
], ListReport.prototype, "defaultPath", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName)
        }
    })
], ListReport.prototype, "defaultTemplateAnnotationPath", void 0);
__decorate([
    (0, decorators_1.childId)('header')
], ListReport.prototype, "header", void 0);
__decorate([
    (0, decorators_1.childId)('fe::FilterBar')
], ListReport.prototype, "filterBar", void 0);
exports.ListReport = ListReport = __decorate([
    (0, decorators_1.pageLayoutInformation)({
        id: 'sap.suite.ui.generic.template.ListReport.view.ListReport',
        pageId: 'sap.suite.ui.generic.template.ListReport'
    })
], ListReport);
