// @sap/ux-specification/dist/src/sync/v4/export/pages/ObjectPage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectPage = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const decorators_1 = require("../../../common/decoration/decorators");
const application_1 = require("../../application");
let ObjectPage = class ObjectPage {
};
exports.ObjectPage = ObjectPage;
__decorate([
    (0, decorators_1.childId)('header')
], ObjectPage.prototype, "header", void 0);
__decorate([
    (0, decorators_1.childId)('layout')
], ObjectPage.prototype, "layout", void 0);
__decorate([
    (0, decorators_1.syncRule)({
        manifest: {
            path: (pageName) => (0, application_1.getSettingsPathForPage)(pageName)
        }
    }),
    (0, decorators_1.validity)({
        since: '1.84.0'
    })
], ObjectPage.prototype, "variantManagement", void 0);
__decorate([
    (0, decorators_1.childId)('sections')
], ObjectPage.prototype, "sections", void 0);
exports.ObjectPage = ObjectPage = __decorate([
    (0, decorators_1.pageLayoutInformation)({
        id: `${ux_specification_types_1.v4.FE_TEMPLATE_V4_OBJECT_PAGE}.view.Details`,
        pageId: ux_specification_types_1.v4.FE_TEMPLATE_V4_OBJECT_PAGE
    })
], ObjectPage);
