// @sap-ux/btp-utils/dist/destination.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isS4HC = exports.getDisplayName = exports.isHTML5DynamicConfigured = exports.isOnPremiseDestination = exports.isFullUrlDestination = exports.isPartialUrlDestination = exports.isGenericODataDestination = exports.isAbapEnvironmentOnBtp = exports.isAbapSystem = exports.DestinationProxyType = exports.WebIDEAdditionalData = exports.WebIDEUsage = exports.ProxyType = exports.Suffix = exports.Authentication = void 0;
/**
 * Support destination authentication types
 */
var Authentication;
(function (Authentication) {
    Authentication["BASIC_AUTHENTICATION"] = "BasicAuthentication";
    Authentication["CLIENT_CERT_AUTHENTICATION"] = "ClientCertificateAuthentication";
    Authentication["NO_AUTHENTICATION"] = "NoAuthentication";
    Authentication["OAUTH2_CLIENT_CREDENTIALS"] = "OAuth2ClientCredentials";
    Authentication["OAUTH2_JWT_BEARER"] = "OAuth2JWTBearer";
    Authentication["OAUTH2_PASSWORD"] = "OAuth2Password";
    Authentication["OAUTH2_REFRESH_TOKEN"] = "OAuth2RefreshToken";
    Authentication["OAUTH2_SAML_BEARER_ASSERTION"] = "OAuth2SAMLBearerAssertion";
    Authentication["OAUTH2_USER_TOKEN_EXCHANGE"] = "OAuth2UserTokenExchange";
    Authentication["SAML_ASSERTION"] = "SAMLAssertion";
})(Authentication || (exports.Authentication = Authentication = {}));
/**
 * Relevant values for display extended destination properties to the UI
 */
var Suffix;
(function (Suffix) {
    Suffix["S4HC"] = "S4HC";
    Suffix["BTP"] = "BTP";
})(Suffix || (exports.Suffix = Suffix = {}));
var ProxyType;
(function (ProxyType) {
    ProxyType["ON_PREMISE"] = "OnPremise";
    ProxyType["INTERNET"] = "Internet";
    ProxyType["PRIVATE_LINK"] = "PrivateLink";
})(ProxyType || (exports.ProxyType = ProxyType = {}));
/**
 * Relevant values for the WebIDEUsage property used to identify different types of destinations.
 */
var WebIDEUsage;
(function (WebIDEUsage) {
    WebIDEUsage["ODATA_GENERIC"] = "odata_gen";
    WebIDEUsage["ODATA_ABAP"] = "odata_abap";
    WebIDEUsage["DEV_ABAP"] = "dev_abap";
    WebIDEUsage["ABAP_CLOUD"] = "abap_cloud";
})(WebIDEUsage || (exports.WebIDEUsage = WebIDEUsage = {}));
/**
 * Possible values for the WebIDEAdditionalData property
 */
var WebIDEAdditionalData;
(function (WebIDEAdditionalData) {
    WebIDEAdditionalData["FULL_URL"] = "full_url";
    WebIDEAdditionalData["API_MGMT"] = "api_mgmt";
})(WebIDEAdditionalData || (exports.WebIDEAdditionalData = WebIDEAdditionalData = {}));
/**
 * Relevant values for DestinationProxyType property
 */
var DestinationProxyType;
(function (DestinationProxyType) {
    DestinationProxyType["ON_PREMISE"] = "OnPremise";
    DestinationProxyType["INTERNET"] = "Internet";
    DestinationProxyType["PRIVATE_LINK"] = "PrivateLink";
})(DestinationProxyType || (exports.DestinationProxyType = DestinationProxyType = {}));
/**
 * Checks whether the provided destination is configured to point to an ABAP system (both cloud and on-premise).
 *
 * @param destination destination info
 * @returns true of the destination is configured for an ABAP system
 */
function isAbapSystem(destination) {
    var _a, _b;
    return Boolean(((_a = destination.WebIDEUsage) === null || _a === void 0 ? void 0 : _a.includes('abap')) ||
        destination['sap-client'] ||
        ((_b = destination['sap-platform']) === null || _b === void 0 ? void 0 : _b.toLocaleLowerCase()) === 'abap');
}
exports.isAbapSystem = isAbapSystem;
/**
 * Checks whether the provided destination is configured to point to an ABAP environment on BTP (aka steampunk).
 *
 * @param destination destination info
 * @returns true of the destination is configured for a steampunk system
 */
function isAbapEnvironmentOnBtp(destination) {
    var _a, _b;
    return Boolean(((_a = destination.WebIDEUsage) === null || _a === void 0 ? void 0 : _a.includes(WebIDEUsage.ABAP_CLOUD)) ||
        ((_b = destination['sap-platform']) === null || _b === void 0 ? void 0 : _b.toLocaleLowerCase()) === 'abap');
}
exports.isAbapEnvironmentOnBtp = isAbapEnvironmentOnBtp;
/**
 * Checks if the destination attributes WebIDEUsage is configured with odata_gen, and not odata_abap
 * We want to maintain existing destinations, perhaps accidentally configured with both, continue
 * working as abap catalogs.
 *
 * @param destination destination info
 * @returns true, if this destination has the the 'odata_gen', and not the 'odata_abap' attribute set
 */
function isGenericODataDestination(destination) {
    var _a;
    return Boolean(((_a = destination.WebIDEUsage) === null || _a === void 0 ? void 0 : _a.includes(WebIDEUsage.ODATA_GENERIC)) &&
        !destination.WebIDEUsage.includes(WebIDEUsage.ODATA_ABAP));
}
exports.isGenericODataDestination = isGenericODataDestination;
/**
 * A destination configured as generic odata without the 'full_url' attribute are considered partial url destinations.
 *
 * @param destination destination info
 * @returns true, if this destination is generic odata and 'full_url' attribute is not set
 */
function isPartialUrlDestination(destination) {
    var _a;
    return Boolean(!((_a = destination.WebIDEAdditionalData) === null || _a === void 0 ? void 0 : _a.includes(WebIDEAdditionalData.FULL_URL)) &&
        isGenericODataDestination(destination));
}
exports.isPartialUrlDestination = isPartialUrlDestination;
/**
 * A destination configured as generic odata WITH the 'full_url' attribute are considered full url destinations.
 *
 * @param destination destination info
 * @returns true, if this destination has the generic odata and 'full_url' attribute is set
 */
function isFullUrlDestination(destination) {
    var _a;
    return Boolean(((_a = destination.WebIDEAdditionalData) === null || _a === void 0 ? void 0 : _a.includes(WebIDEAdditionalData.FULL_URL)) &&
        isGenericODataDestination(destination));
}
exports.isFullUrlDestination = isFullUrlDestination;
/**
 * Checks whether the provided destination is configured as an on-premise system.
 *
 * @param destination destination info
 * @returns true if the destination is configured as an on-premise system
 */
function isOnPremiseDestination(destination) {
    return Boolean(destination.ProxyType.includes(DestinationProxyType.ON_PREMISE));
}
exports.isOnPremiseDestination = isOnPremiseDestination;
/**
 * Validate if a destination has the property `HTML5.DynamicDestination` configured.
 *
 * @param destination destination info
 * @returns true, if this destination has HTML5.DynamicDestination configured
 */
function isHTML5DynamicConfigured(destination) {
    return Boolean(destination['HTML5.DynamicDestination']);
}
exports.isHTML5DynamicConfigured = isHTML5DynamicConfigured;
/**
 * Escape any special RegExp character that we want to use literally.
 *
 * @param str string input
 * @returns string a cleansed version of the input
 */
function escapeRegExp(str) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
}
/**
 * Trim, cleanse and return a system name appended with the appropriate suffix i.e. BTP | S4HC.
 *
 * @param destinationName name of the destination
 * @param suffix the appropriate suffix appended, BTP | S4HC
 * @returns string return an escaped string, appended with the appropriate suffix
 */
function addSuffix(destinationName, suffix) {
    const suffixStr = ` (${suffix})`;
    return RegExp(`${escapeRegExp(suffixStr)}$`).exec(destinationName.trim())
        ? destinationName
        : `${destinationName} (${suffix})`;
}
/**
 * Generate a display name using the destination name, username if supplied and the appropriate suffix i.e. BTP | S4HC.
 *
 * @param destination destination info
 * @param displayUsername name to display with destination
 * @returns string a newly generated string value with the name of the destination, username if present and the system type
 */
function getDisplayName(destination, displayUsername) {
    const userDisplayName = displayUsername ? ` [${displayUsername}]` : '';
    let systemDisplayName;
    if (isAbapEnvironmentOnBtp(destination)) {
        systemDisplayName = addSuffix(destination.Name, Suffix.BTP);
    }
    else if (isS4HC(destination)) {
        systemDisplayName = addSuffix(destination.Name, Suffix.S4HC);
    }
    else {
        systemDisplayName = destination.Name;
    }
    return `${systemDisplayName}${userDisplayName}`;
}
exports.getDisplayName = getDisplayName;
/**
 * Checks whether the provided destination is configured to point to an S/4 HANA system.
 *
 * @param destination destination info
 * @returns boolean if the destination is configured for an SAP S/4HANA system
 */
function isS4HC(destination) {
    var _a;
    return Boolean(((_a = destination.WebIDEUsage) === null || _a === void 0 ? void 0 : _a.includes(WebIDEUsage.ODATA_ABAP)) &&
        destination.Authentication === Authentication.SAML_ASSERTION &&
        destination.ProxyType === ProxyType.INTERNET);
}
exports.isS4HC = isS4HC;
