// @sap-ux/fe-fpm-writer/dist/common/file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extendJSON = exports.getJsonSpace = exports.detectTabSpacing = void 0;
const CHAR_SPACE = ' ';
const CHAR_TAB = '\t';
/**
 * Method returns tab info for passed line.
 *
 * @param line - line with tab spacing
 * @returns tab size information
 */
function getLineTabInfo(line) {
    let tabSize;
    const symbol = line[0] === CHAR_TAB ? CHAR_TAB : CHAR_SPACE;
    // get count of tabs
    for (let i = 0; i < line.length; i++) {
        const char = line[i];
        if (char !== symbol) {
            tabSize = {
                size: i,
                useTabSymbol: symbol === CHAR_TAB
            };
            break;
        }
    }
    return tabSize;
}
/**
 * Method calculates tab space info for passed file content.
 *
 * @param content - file content.
 * @returns tab size information.
 */
function detectTabSpacing(content) {
    let tabSize;
    const tabSymbols = [CHAR_SPACE, CHAR_TAB];
    const lines = content.split(/\r\n|\n/);
    const lineWithSpacing = lines.find((line) => {
        return tabSymbols.includes(line[0]);
    });
    if (lineWithSpacing) {
        tabSize = getLineTabInfo(lineWithSpacing);
    }
    return tabSize;
}
exports.detectTabSpacing = detectTabSpacing;
/**
 * Method calculates tab spacing parameter for 'JSON.stringify' method.
 *
 * @param fs - the mem-fs editor instance.
 * @param filePath - path to file to read.
 * @param tabInfo - External tab configuration.
 * @returns tab size information.
 */
function getJsonSpace(fs, filePath, tabInfo) {
    if (!tabInfo) {
        // 'tabInfo'  was not passed - calculate 'tabInfo' by checking existing content of target file
        const content = fs.read(filePath);
        tabInfo = detectTabSpacing(content);
    }
    let space;
    if (tabInfo) {
        // 'tabInfo' exists - it was passed as custom configuration or calculated from target file
        if (tabInfo.useTabSymbol) {
            // Tab symbol should be used as tab
            space = CHAR_TAB.repeat(tabInfo.size || 1);
        }
        else {
            // Spaces should be used as tab
            space = tabInfo.size;
        }
    }
    return space;
}
exports.getJsonSpace = getJsonSpace;
/**
 * Method extends target JSON file with passed JSOn content.
 * Method uses 'fs.extendJSON', but applies additional calculation to reuse existing content tab sizing information.
 *
 * @param fs - the mem-fs editor instance.
 * @param params - options for JSON extend.
 */
function extendJSON(fs, params) {
    const { filepath, content, replacer } = params;
    const space = getJsonSpace(fs, filepath, params.tabInfo);
    // Write json
    fs.extendJSON(filepath, JSON.parse(content), replacer, space);
}
exports.extendJSON = extendJSON;
