// @sap/ux-specification/dist/src/sync/v2/application.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppSettings = exports.ApplicationV2 = exports.getObjectPageCustomSectionPath = exports.getViewExtensionsPath = exports.getChartSettingsOBJsectionPathV2 = exports.getTableSettingsOBJsectionPathV2 = exports.getAppSettingOBJsectionPathV2 = exports.getGlobalTableSettingsPathV2 = exports.getAppSettingOBJPathV2 = exports.getAppRootSettingOBJPath = exports.getKPISettingsPath = exports.getGlobalSettingsPath = exports.getCardSettingsPath = exports.getOVPsettingPath = exports.getCreationFieldsPathV2 = exports.getDataLoadSettingPathV2 = exports.getChartSettingPathV2 = exports.getTableSettingPathV2 = exports.getAppSettingPathV2 = void 0;
const decoration_1 = require("../common/decoration");
const ux_specification_types_1 = require("@sap/ux-specification-types");
const getAppSettingPathV2 = (pageKeys) => {
    return `${ux_specification_types_1.ManifestSection.generic}/pages/${pageKeys[0]}/component/settings`;
};
exports.getAppSettingPathV2 = getAppSettingPathV2;
const getTableSettingPathV2 = (pageKeys) => {
    return (0, exports.getAppSettingPathV2)(pageKeys) + '/tableSettings';
};
exports.getTableSettingPathV2 = getTableSettingPathV2;
const getChartSettingPathV2 = (pageKeys) => {
    return (0, exports.getAppSettingPathV2)(pageKeys) + '/chartSettings';
};
exports.getChartSettingPathV2 = getChartSettingPathV2;
const getDataLoadSettingPathV2 = (pageKeys) => {
    return (0, exports.getAppSettingPathV2)(pageKeys) + '/dataLoadSettings';
};
exports.getDataLoadSettingPathV2 = getDataLoadSettingPathV2;
const getCreationFieldsPathV2 = (pageKeys, key) => {
    return `${(0, exports.getTableSettingPathV2)(pageKeys)}/createWithParameterDialog/fields/${key}`;
};
exports.getCreationFieldsPathV2 = getCreationFieldsPathV2;
const getOVPsettingPath = () => {
    return ux_specification_types_1.ManifestSection.ovp;
};
exports.getOVPsettingPath = getOVPsettingPath;
const getCardSettingsPath = (cardKeys) => {
    const pathParts = [ux_specification_types_1.ManifestSection.ovp + '/cards'];
    cardKeys.forEach((block) => {
        if (block in ux_specification_types_1.v2.CardSettingsType) {
            pathParts.push('settings');
        }
        else {
            pathParts.push(block);
        }
    });
    pathParts.pop();
    return pathParts.join('/');
};
exports.getCardSettingsPath = getCardSettingsPath;
const getGlobalSettingsPath = (cardKey) => {
    const pathParts = [ux_specification_types_1.ManifestSection.ovp];
    cardKey.forEach((block) => {
        pathParts.push(block);
    });
    pathParts.pop();
    return pathParts.join('/');
};
exports.getGlobalSettingsPath = getGlobalSettingsPath;
// This corresponds to ALP KPI settings
const getKPISettingsPath = (pageKeys, kpiKey) => {
    return `${ux_specification_types_1.ManifestSection.generic}/pages/${pageKeys[0]}/component/settings/keyPerformanceIndicators/${kpiKey}`;
};
exports.getKPISettingsPath = getKPISettingsPath;
// This corresponds to List Report and Object page global settings
const getAppRootSettingOBJPath = (pageKeys) => {
    const pathParts = [ux_specification_types_1.ManifestSection.generic];
    pageKeys.forEach((pageKey) => {
        pathParts.push('pages/' + pageKey);
    });
    return pathParts.join('/');
};
exports.getAppRootSettingOBJPath = getAppRootSettingOBJPath;
// This corresponds to global object page settings
const getAppSettingOBJPathV2 = (pageKeys) => {
    const pathParts = [];
    pathParts.push((0, exports.getAppRootSettingOBJPath)(pageKeys));
    pathParts.push('component/settings');
    return pathParts.join('/');
};
exports.getAppSettingOBJPathV2 = getAppSettingOBJPathV2;
const getGlobalTableSettingsPathV2 = (pageKeys) => {
    const path = (0, exports.getAppSettingOBJPathV2)(pageKeys);
    return path + '/tableSettings';
};
exports.getGlobalTableSettingsPathV2 = getGlobalTableSettingsPathV2;
const getAppSettingOBJsectionPathV2 = (pageKeys, _manifest, sectionId) => {
    const pathParts = [];
    pathParts.push((0, exports.getAppRootSettingOBJPath)(pageKeys));
    const manifestSectionId = sectionId.replace(ux_specification_types_1.VOCWITHCOLONS, '::com.sap.vocabularies');
    pathParts.push('component/settings/sections/' + manifestSectionId);
    return pathParts.join('/');
};
exports.getAppSettingOBJsectionPathV2 = getAppSettingOBJsectionPathV2;
const getTableSettingsOBJsectionPathV2 = (pageKeys, _manifest, sectionId) => {
    const path = (0, exports.getAppSettingOBJsectionPathV2)(pageKeys, {}, sectionId);
    return path + '/tableSettings';
};
exports.getTableSettingsOBJsectionPathV2 = getTableSettingsOBJsectionPathV2;
const getChartSettingsOBJsectionPathV2 = (pageKeys, _manifest, sectionId) => {
    const path = (0, exports.getAppSettingOBJsectionPathV2)(pageKeys, {}, sectionId);
    return path + '/chartSettings';
};
exports.getChartSettingsOBJsectionPathV2 = getChartSettingsOBJsectionPathV2;
const getViewExtensionsPath = () => {
    return `sap.ui5/extends/extensions/${ux_specification_types_1.v2.SAPUI5_VIEW_EXTENSION}/${ux_specification_types_1.v2.SAPUI5_VIEW_EXTENSION_OBJECT_PAGE}`;
};
exports.getViewExtensionsPath = getViewExtensionsPath;
const getObjectPageCustomSectionPath = (pageName, _manifest, sectionId, ...pathParts) => {
    const separator = '|';
    const sectionIdParts = sectionId.split(separator);
    const pageNameParts = pageName[pageName.length - 1].split(separator);
    let idLastPart = '';
    if (sectionIdParts.length === 4) {
        // Four parts ==> the last one is incremented to avoid getting the same path
        // We need to merge 3 and 4
        idLastPart = [sectionIdParts[2], sectionIdParts[3]].join(separator);
    }
    else if (sectionIdParts.length === 3) {
        idLastPart = sectionIdParts[2];
    }
    const customSectionId = [sectionIdParts[0], pageNameParts[pageNameParts.length - 1], idLastPart].join(separator);
    // Sub-path parts are passed
    pathParts.unshift(customSectionId);
    return `${(0, exports.getViewExtensionsPath)()}/${pathParts.join('/')}`;
};
exports.getObjectPageCustomSectionPath = getObjectPageCustomSectionPath;
class ApplicationV2 {
}
exports.ApplicationV2 = ApplicationV2;
__decorate([
    (0, decoration_1.childId)('settings')
], ApplicationV2.prototype, "settings", void 0);
class AppSettings {
}
exports.AppSettings = AppSettings;
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: () => 'sap.app'
        }
    })
], AppSettings.prototype, "title", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: () => 'sap.app'
        }
    })
], AppSettings.prototype, "description", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: () => ux_specification_types_1.GENERICAPPSETTINGS
        }
    }),
    (0, decoration_1.validity)({
        since: '1.50.0'
    })
], AppSettings.prototype, "flexibleColumnLayout", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: () => ux_specification_types_1.GENERICAPPSETTINGS
        }
    }),
    (0, decoration_1.validity)({
        since: '1.46.0'
    })
], AppSettings.prototype, "forceGlobalRefresh", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: () => ux_specification_types_1.GENERICAPPSETTINGS
        }
    }),
    (0, decoration_1.validity)({
        since: '1.74.0'
    })
], AppSettings.prototype, "tableColumnVerticalAlignment", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: () => 'sap.ui5'
        }
    }),
    (0, decoration_1.validity)({
        since: '1.56.0'
    })
], AppSettings.prototype, "flexEnabled", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: () => ux_specification_types_1.GENERICAPPSETTINGS
        }
    }),
    (0, decoration_1.validity)({
        since: '1.96.0'
    })
], AppSettings.prototype, "draftDiscardConfirmationSettings", void 0);
__decorate([
    (0, decoration_1.syncRule)({
        manifest: {
            path: () => ux_specification_types_1.GENERICAPPSETTINGS
        }
    }),
    (0, decoration_1.validity)({
        since: '1.96.0'
    })
], AppSettings.prototype, "statePreservationMode", void 0);
