// @sap/ux-specification/dist/src/sync/v2/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionTypeV2 = exports.SectionTypeV2 = exports.OP_STANDARD_FOOTER_ACTION_PREFIX = exports.OP_STANDARD_HEADER_ACTION_PREFIX = exports.LINEITEMTABLETYPES = exports.GRIDTABLE = exports.ANALYTICALTABLE = exports.TREETABLE = exports.INLINEDELETETABLE = exports.MULTISELECTTABLE = void 0;
exports.MULTISELECTTABLE = 'ResponsiveTableWithMultiSelect<LineItems>';
exports.INLINEDELETETABLE = 'ResponsiveTableWithInlineDelete<LineItems>';
exports.TREETABLE = 'TreeTable<LineItems>';
exports.ANALYTICALTABLE = 'AnalyticalTable<LineItems>';
exports.GRIDTABLE = 'GridTable<LineItems>';
exports.LINEITEMTABLETYPES = [exports.MULTISELECTTABLE, exports.INLINEDELETETABLE, exports.TREETABLE, exports.ANALYTICALTABLE, exports.GRIDTABLE];
exports.OP_STANDARD_HEADER_ACTION_PREFIX = 'template::';
exports.OP_STANDARD_FOOTER_ACTION_PREFIX = 'template:::ObjectPageAction:::';
var SectionTypeV2;
(function (SectionTypeV2) {
    SectionTypeV2["SectionAddress"] = "ObjectPageSectionAddressV2";
    SectionTypeV2["SectionChart"] = "ObjectPageSectionChartV2";
    SectionTypeV2["SectionContact"] = "ObjectPageSectionContactV2";
    SectionTypeV2["SectionForm"] = "ObjectPageSectionFormV2";
    SectionTypeV2["Section"] = "ObjectPageSectionV2";
    SectionTypeV2["SubSection"] = "ObjectPageSubSectionV2";
    SectionTypeV2["SectionTable"] = "ObjectPageSectionTableV2";
    SectionTypeV2["SubSectionForm"] = "ObjectPageSubSectionFormV2";
    SectionTypeV2["SubSectionTable"] = "ObjectPageSubSectionTableV2";
    SectionTypeV2["SubSectionAddress"] = "ObjectPageSubSectionAddressV2";
    SectionTypeV2["SubSectionContact"] = "ObjectPageSubSectionContactV2";
    SectionTypeV2["SubSectionChart"] = "ObjectPageSubSectionChartV2";
    SectionTypeV2["CommonHeaderFacetSettings"] = "CommonHeaderFacetSettings";
    SectionTypeV2["HeaderSection"] = "ObjectPageHeaderSection";
    SectionTypeV2["HeaderSectionAddress"] = "ObjectPageHeaderSectionAddress";
    SectionTypeV2["HeaderSectionChart"] = "ObjectPageHeaderSectionChart";
    SectionTypeV2["HeaderSectionCustomFragment"] = "ObjectPageCustomHeaderSectionFragment";
    SectionTypeV2["HeaderSectionDataPoint"] = "ObjectPageHeaderSectionDataPoint";
    SectionTypeV2["HeaderSectionForm"] = "ObjectPageHeaderSectionForm";
})(SectionTypeV2 || (exports.SectionTypeV2 = SectionTypeV2 = {}));
var ActionTypeV2;
(function (ActionTypeV2) {
    ActionTypeV2["Form"] = "FormAction";
    ActionTypeV2["OPHeader"] = "ObjectPageHeaderAction";
    ActionTypeV2["OPToolBar"] = "ObjectPageToolBarAction";
    ActionTypeV2["OPFooter"] = "ObjectPageFooterAction";
    ActionTypeV2["OPColumn"] = "ObjectPageTableColumnAction";
    ActionTypeV2["ToolBar"] = "ToolBarAction";
    ActionTypeV2["Footer"] = "FooterAction";
    ActionTypeV2["Column"] = "TableColumnAction";
    ActionTypeV2["ToolBarIntentBased"] = "ToolBarActionIntentBased";
    ActionTypeV2["FooterIntentBased"] = "FooterActionIntentBased";
    ActionTypeV2["ColumnIntentBased"] = "TableColumnActionIntentBased";
    ActionTypeV2["HeaderStandard"] = "HeaderActionStandard";
    ActionTypeV2["OPHeaderStandard"] = "ObjectPageHeaderActionStandard";
    ActionTypeV2["OPFooterStandard"] = "ObjectPageFooterActionStandard";
})(ActionTypeV2 || (exports.ActionTypeV2 = ActionTypeV2 = {}));
