// @sap-ux/vocabularies-types/vocabularies/Analytics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnalyticsAnnotationTypes = exports.AnalyticsAnnotationTerms = void 0;
var AnalyticsAnnotationTerms;
(function (AnalyticsAnnotationTerms) {
    AnalyticsAnnotationTerms["Dimension"] = "com.sap.vocabularies.Analytics.v1.Dimension";
    AnalyticsAnnotationTerms["Measure"] = "com.sap.vocabularies.Analytics.v1.Measure";
    AnalyticsAnnotationTerms["AccumulativeMeasure"] = "com.sap.vocabularies.Analytics.v1.AccumulativeMeasure";
    AnalyticsAnnotationTerms["RolledUpPropertyCount"] = "com.sap.vocabularies.Analytics.v1.RolledUpPropertyCount";
    AnalyticsAnnotationTerms["DrillURL"] = "com.sap.vocabularies.Analytics.v1.DrillURL";
    AnalyticsAnnotationTerms["PlanningAction"] = "com.sap.vocabularies.Analytics.v1.PlanningAction";
    AnalyticsAnnotationTerms["AggregatedProperties"] = "com.sap.vocabularies.Analytics.v1.AggregatedProperties";
    AnalyticsAnnotationTerms["AggregatedProperty"] = "com.sap.vocabularies.Analytics.v1.AggregatedProperty";
    AnalyticsAnnotationTerms["AnalyticalContext"] = "com.sap.vocabularies.Analytics.v1.AnalyticalContext";
})(AnalyticsAnnotationTerms = exports.AnalyticsAnnotationTerms || (exports.AnalyticsAnnotationTerms = {}));
var AnalyticsAnnotationTypes;
(function (AnalyticsAnnotationTypes) {
    AnalyticsAnnotationTypes["AggregatedPropertyType"] = "com.sap.vocabularies.Analytics.v1.AggregatedPropertyType";
    AnalyticsAnnotationTypes["AnalyticalContextType"] = "com.sap.vocabularies.Analytics.v1.AnalyticalContextType";
})(AnalyticsAnnotationTypes = exports.AnalyticsAnnotationTypes || (exports.AnalyticsAnnotationTypes = {}));
