// @sap-ux/vocabularies-types/vocabularies/CodeList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeListAnnotationTypes = exports.CodeListAnnotationTerms = void 0;
var CodeListAnnotationTerms;
(function (CodeListAnnotationTerms) {
    CodeListAnnotationTerms["CurrencyCodes"] = "com.sap.vocabularies.CodeList.v1.CurrencyCodes";
    CodeListAnnotationTerms["UnitsOfMeasure"] = "com.sap.vocabularies.CodeList.v1.UnitsOfMeasure";
    CodeListAnnotationTerms["StandardCode"] = "com.sap.vocabularies.CodeList.v1.StandardCode";
    CodeListAnnotationTerms["ExternalCode"] = "com.sap.vocabularies.CodeList.v1.ExternalCode";
    CodeListAnnotationTerms["IsConfigurationDeprecationCode"] = "com.sap.vocabularies.CodeList.v1.IsConfigurationDeprecationCode";
})(CodeListAnnotationTerms = exports.CodeListAnnotationTerms || (exports.CodeListAnnotationTerms = {}));
var CodeListAnnotationTypes;
(function (CodeListAnnotationTypes) {
    CodeListAnnotationTypes["CodeListSource"] = "com.sap.vocabularies.CodeList.v1.CodeListSource";
})(CodeListAnnotationTypes = exports.CodeListAnnotationTypes || (exports.CodeListAnnotationTypes = {}));
