// @sap-ux/vocabularies-types/vocabularies/Hierarchy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HierarchyAnnotationTypes = exports.HierarchyAnnotationTerms = void 0;
var HierarchyAnnotationTerms;
(function (HierarchyAnnotationTerms) {
    HierarchyAnnotationTerms["RecursiveHierarchy"] = "com.sap.vocabularies.Hierarchy.v1.RecursiveHierarchy";
    HierarchyAnnotationTerms["RecursiveHierarchyActions"] = "com.sap.vocabularies.Hierarchy.v1.RecursiveHierarchyActions";
    HierarchyAnnotationTerms["MatchCount"] = "com.sap.vocabularies.Hierarchy.v1.MatchCount";
    HierarchyAnnotationTerms["RecursiveHierarchySupported"] = "com.sap.vocabularies.Hierarchy.v1.RecursiveHierarchySupported";
})(HierarchyAnnotationTerms = exports.HierarchyAnnotationTerms || (exports.HierarchyAnnotationTerms = {}));
var HierarchyAnnotationTypes;
(function (HierarchyAnnotationTypes) {
    HierarchyAnnotationTypes["RecursiveHierarchyType"] = "com.sap.vocabularies.Hierarchy.v1.RecursiveHierarchyType";
    HierarchyAnnotationTypes["RecursiveHierarchyActionsType"] = "com.sap.vocabularies.Hierarchy.v1.RecursiveHierarchyActionsType";
    HierarchyAnnotationTypes["TopLevelsExpandType"] = "com.sap.vocabularies.Hierarchy.v1.TopLevelsExpandType";
})(HierarchyAnnotationTypes = exports.HierarchyAnnotationTypes || (exports.HierarchyAnnotationTypes = {}));
