// @sap-ux/vocabularies-types/vocabularies/PersonalData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PersonalDataAnnotationTypes = exports.PersonalDataAnnotationTerms = void 0;
var PersonalDataAnnotationTerms;
(function (PersonalDataAnnotationTerms) {
    PersonalDataAnnotationTerms["EntitySemantics"] = "com.sap.vocabularies.PersonalData.v1.EntitySemantics";
    PersonalDataAnnotationTerms["DataSubjectRole"] = "com.sap.vocabularies.PersonalData.v1.DataSubjectRole";
    PersonalDataAnnotationTerms["DataSubjectRoleDescription"] = "com.sap.vocabularies.PersonalData.v1.DataSubjectRoleDescription";
    PersonalDataAnnotationTerms["FieldSemantics"] = "com.sap.vocabularies.PersonalData.v1.FieldSemantics";
    PersonalDataAnnotationTerms["IsPotentiallyPersonal"] = "com.sap.vocabularies.PersonalData.v1.IsPotentiallyPersonal";
    PersonalDataAnnotationTerms["IsPotentiallySensitive"] = "com.sap.vocabularies.PersonalData.v1.IsPotentiallySensitive";
})(PersonalDataAnnotationTerms = exports.PersonalDataAnnotationTerms || (exports.PersonalDataAnnotationTerms = {}));
var PersonalDataAnnotationTypes;
(function (PersonalDataAnnotationTypes) {
})(PersonalDataAnnotationTypes = exports.PersonalDataAnnotationTypes || (exports.PersonalDataAnnotationTypes = {}));
