// @sap-ux/vocabularies-types/vocabularies/Validation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationAnnotationTypes = exports.ValidationAnnotationTerms = void 0;
var ValidationAnnotationTerms;
(function (ValidationAnnotationTerms) {
    ValidationAnnotationTerms["Pattern"] = "Org.OData.Validation.V1.Pattern";
    ValidationAnnotationTerms["Minimum"] = "Org.OData.Validation.V1.Minimum";
    ValidationAnnotationTerms["Maximum"] = "Org.OData.Validation.V1.Maximum";
    ValidationAnnotationTerms["Exclusive"] = "Org.OData.Validation.V1.Exclusive";
    ValidationAnnotationTerms["AllowedValues"] = "Org.OData.Validation.V1.AllowedValues";
    ValidationAnnotationTerms["MultipleOf"] = "Org.OData.Validation.V1.MultipleOf";
    ValidationAnnotationTerms["Constraint"] = "Org.OData.Validation.V1.Constraint";
    ValidationAnnotationTerms["ItemsOf"] = "Org.OData.Validation.V1.ItemsOf";
    ValidationAnnotationTerms["OpenPropertyTypeConstraint"] = "Org.OData.Validation.V1.OpenPropertyTypeConstraint";
    ValidationAnnotationTerms["DerivedTypeConstraint"] = "Org.OData.Validation.V1.DerivedTypeConstraint";
    ValidationAnnotationTerms["AllowedTerms"] = "Org.OData.Validation.V1.AllowedTerms";
    ValidationAnnotationTerms["ApplicableTerms"] = "Org.OData.Validation.V1.ApplicableTerms";
    ValidationAnnotationTerms["MaxItems"] = "Org.OData.Validation.V1.MaxItems";
    ValidationAnnotationTerms["MinItems"] = "Org.OData.Validation.V1.MinItems";
})(ValidationAnnotationTerms = exports.ValidationAnnotationTerms || (exports.ValidationAnnotationTerms = {}));
var ValidationAnnotationTypes;
(function (ValidationAnnotationTypes) {
    ValidationAnnotationTypes["AllowedValue"] = "Org.OData.Validation.V1.AllowedValue";
    ValidationAnnotationTypes["ConstraintType"] = "Org.OData.Validation.V1.ConstraintType";
    ValidationAnnotationTypes["ItemsOfType"] = "Org.OData.Validation.V1.ItemsOfType";
})(ValidationAnnotationTypes = exports.ValidationAnnotationTypes || (exports.ValidationAnnotationTypes = {}));
