// @sap/ux-specification/dist/src/sync/v4/import/fragment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readFilterFieldsExtensions = exports.readColumnExtensions = exports.readSectionExtensions = exports.readHeaderFacetExtensions = void 0;
const ux_specification_types_1 = require("@sap/ux-specification-types");
const path_1 = require("path");
/**
 * Determines a related fragment file and returns its content
 */
function readFragmentFile(fragmentName, fragmentFiles, manifestPart, appId) {
    let fragmentData;
    if (fragmentName) {
        const fragmentId = fragmentName.substring(appId.length + 1);
        const pathParts = fragmentId.split('.');
        pathParts[pathParts.length - 1] = pathParts[pathParts.length - 1] + '.fragment.xml';
        const extensionPath = (0, path_1.join)('webapp', ...pathParts);
        for (const file of fragmentFiles) {
            if (file.dataSourceUri.endsWith(extensionPath)) {
                fragmentData = {
                    manifestPart,
                    fragmentFileContent: file.fileContent
                };
                break;
            }
        }
    }
    return fragmentData;
}
/**
 * Finds fragment information in a control configuration segment
 * @param segment - Segment of the control configuration
 * @param fragmentFiles - list of all fragment files' content
 * @param appId - Application ID
 * @returns a list of fragment data from the given segment
 */
function findFragmentInSegment(segment, fragmentFiles, appId) {
    const fragmentData = [];
    for (const key in segment) {
        //extension as part of controlConfiguration
        const fragment = readFragmentFile(segment[key]['template'], fragmentFiles, segment[key], appId);
        if (fragment) {
            fragmentData.push(fragment);
        }
    }
    return fragmentData;
}
/**
 * Reads an extension from a given segment of a V4 target's controlConfiguration
 */
function readControlConfigurationExtension(manifest, target, targetSegment, appId, fragmentFiles) {
    const targets = manifest[ux_specification_types_1.ManifestSection.ui5].routing.targets;
    let fragmentData = [];
    for (const pageKey in targets) {
        if (pageKey === target) {
            const controlConfiguration = targets[pageKey].options?.settings?.['controlConfiguration'];
            for (const configKey in controlConfiguration) {
                const dataOfSegment = findFragmentInSegment(controlConfiguration[configKey][targetSegment], fragmentFiles, appId);
                fragmentData = [...dataOfSegment];
            }
        }
    }
    return fragmentData;
}
/**
 * Note: header or footer actions are not linked to a fragment, thus cannot be found here
 */
/**
 * Reads the header facet extensions
 */
function readHeaderFacetExtensions(manifest, target, appId, fragmentFiles) {
    const fragmentData = [];
    const targets = manifest[ux_specification_types_1.ManifestSection.ui5].routing.targets;
    for (const pageKey in targets) {
        if (pageKey === target) {
            const targetSettings = targets[pageKey].options?.settings;
            // Read extensions for header facets
            const configBasedFacets = targetSettings?.controlConfiguration?.[`@${"com.sap.vocabularies.UI.v1.HeaderFacets" /* UIAnnotationTerms.HeaderFacets */}`]?.facets;
            const contentBasedFacets = targetSettings?.content?.header?.facets;
            const headerFacets = Object.assign({}, configBasedFacets, contentBasedFacets);
            for (const key in headerFacets) {
                const fragment = readFragmentFile(headerFacets[key]['name'], fragmentFiles, headerFacets[key], appId);
                if (fragment) {
                    fragmentData.push(fragment);
                }
            }
        }
    }
    return fragmentData;
}
exports.readHeaderFacetExtensions = readHeaderFacetExtensions;
/**
 * Reads the section extensions (merge content with control configuration, content-based definitions win)
 */
function readSectionExtensions(manifest, target, appId, fragmentFiles) {
    const fragmentData = [];
    const targets = manifest[ux_specification_types_1.ManifestSection.ui5].routing.targets;
    for (const pageKey in targets) {
        if (pageKey === target) {
            const targetSettings = targets[pageKey].options?.settings;
            // Read extensions for sections
            const contentBasedSections = targetSettings?.content?.body?.sections;
            const configBasedSections = targetSettings?.controlConfiguration?.[`@${"com.sap.vocabularies.UI.v1.Facets" /* UIAnnotationTerms.Facets */}`]?.sections;
            const sections = Object.assign({}, configBasedSections, contentBasedSections);
            for (const key in sections) {
                const fragment = readFragmentFile(sections[key]['name'], fragmentFiles, sections[key], appId);
                if (fragment) {
                    fragmentData.push(fragment);
                }
            }
        }
    }
    return fragmentData;
}
exports.readSectionExtensions = readSectionExtensions;
/**
 * Reads the columns extensions (from control configuration)
 */
function readColumnExtensions(manifest, target, appId, fragmentFiles) {
    return readControlConfigurationExtension(manifest, target, 'columns', appId, fragmentFiles);
}
exports.readColumnExtensions = readColumnExtensions;
/**
 * Reads the filter fields extensions (from control configuration)
 */
function readFilterFieldsExtensions(manifest, target, appId, fragmentFiles) {
    return readControlConfigurationExtension(manifest, target, 'filterFields', appId, fragmentFiles);
}
exports.readFilterFieldsExtensions = readFilterFieldsExtensions;
