// @sap-ux/axios-extension/dist/auth/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UAATimeoutError = exports.TimeoutError = exports.ConnectionError = exports.BaseError = void 0;
/** Wraps `Error`. Used as base class for other specific errors */
class BaseError extends Error {
    /**
     * Constructor taking a message and any object.
     *
     * @param message human readable error message
     * @param cause object causing the issue
     */
    constructor(message, cause) {
        super(message);
        this.cause = cause;
        this.name = this.constructor.name;
    }
}
exports.BaseError = BaseError;
/**
 * Error representing a connection problem.
 */
class ConnectionError extends BaseError {
}
exports.ConnectionError = ConnectionError;
/**
 * Error representing a timeout.
 */
class TimeoutError extends BaseError {
}
exports.TimeoutError = TimeoutError;
/**
 * Error representing a timeout from a UAA service.
 */
class UAATimeoutError extends TimeoutError {
}
exports.UAATimeoutError = UAATimeoutError;
