// @sap/ux-specification/dist/src/sync/v2/export/factory.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataInstanceFactoryV2 = void 0;
const controls = __importStar(require("./controls"));
const pages = __importStar(require("./pages"));
const factory_1 = require("../../common/decoration/factory");
const ux_specification_types_1 = require("@sap/ux-specification-types");
/**
 * Note: If the key of the elements corresponds to the name of the definition in schema, you do not have to list it explicitly here.
 * The following tables only comprise entries for elements that do not follow the official pattern (yet)
 */
const ALPElements = {
//no deviations
};
const LRElements = {
    Table: controls.ResponsiveTable,
    ResponsiveTableWithMultiSelect: controls.ResponsiveTable,
    ResponsiveTableWithInlineDelete: controls.ResponsiveTable
};
const OVPElements = {
    [ux_specification_types_1.v2.CardSettingsType.listCardSettings]: controls.ListCard,
    [ux_specification_types_1.v2.CardSettingsType.listCardSettingsv4]: controls.ListCard,
    [ux_specification_types_1.v2.CardSettingsType.analyticalCardSettings]: controls.AnalyticalCard,
    [ux_specification_types_1.v2.CardSettingsType.analyticalCardSettingsv4]: controls.AnalyticalCard,
    [ux_specification_types_1.v2.CardSettingsType.stackCardSettings]: controls.StackCard,
    [ux_specification_types_1.v2.CardSettingsType.linkListCardSettings]: controls.LinklistCard,
    [ux_specification_types_1.v2.CardSettingsType.tableCardSettings]: controls.TableCard,
    [ux_specification_types_1.v2.CardSettingsType.tableCardSettingsv4]: controls.TableCard
};
const OPElements = {
    [ux_specification_types_1.SAPUI5_VIEW_CLASS]: controls.ObjectPageCustomSectionView,
    [ux_specification_types_1.SAPUI5_FRAGMENT_CLASS]: controls.ObjectPageCustomSectionFragment,
    Table: controls.ObjectPageResponsiveTable, //default
    ObjectPageResponsiveTableWithMultiSelect: controls.ObjectPageResponsiveTable,
    ObjectPageResponsiveTableWithInlineDelete: controls.ObjectPageResponsiveTable
};
const pageTypes = {
    [ux_specification_types_1.PageTypeV2.AnalyticalListPage]: ALPElements,
    [ux_specification_types_1.PageTypeV2.ListReport]: LRElements,
    [ux_specification_types_1.PageTypeV2.ObjectPage]: OPElements,
    [ux_specification_types_1.PageTypeV2.OverviewPage]: OVPElements
};
/**
 * V2 specific factory class for creating instances of (reflect) metadata classes
 */
class MetadataInstanceFactoryV2 extends factory_1.MetadataFactory {
    constructor() {
        super(controls, pages, pageTypes);
    }
}
exports.MetadataInstanceFactoryV2 = MetadataInstanceFactoryV2;
